# Primer proceso: Carga y limpieza de datos

En este primer notebook se hará la carga, limpieza y primera exploración breve sobre los datos a tratar para el EDA.

Se importa la librería 

In [1]:
import pandas as pd
import sys
sys.path.append("../utils")

from funciones import *

### Carga dataset original

Se carga el archivo csv subido previamente en la carpeta data, el cual contiene los datos sobre los que se trabajará

In [2]:
# Cargar datos
df = pd.read_csv("../data/develop_dataset_2025.csv", encoding='utf-8')

C:\Users\maxim\AppData\Local\Temp\ipykernel_34104\4086160946.py:2: DtypeWarning: Columns (0: JobSatPoints_15_TEXT, 1: DatabaseHaveEntry, 2: DevEnvHaveEntry, 3: SOTagsHaveEntry, 4: SOTagsWant Entry, 5: OfficeStackWantEntry, 6: CommPlatformHaveEntr, 7: CommPlatformWantEntr, 8: SO_Actions_15_TEXT, 9: AIAgentOrchestration, 10: AIAgentObsWrite) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("../data/develop_dataset_2025.csv", encoding='utf-8')


### Inspección inicial

In [3]:
# Dimensiones
print(f"Filas: {df.shape[0]:,} | Columnas: {df.shape[1]}")

Filas: 49,191 | Columnas: 172


In [4]:
# Primeros valores
df.head()

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,...,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,Vertex AI,NaN,NaN,NaN,ChatGPT,NaN,When I don’t trust AI’s answers,"Troubleshooting, profiling, debugging",61256.0,10.0
1,2,I am a developer by profession,25-34 years old,"Associate degree (A.A., A.S., etc.)",Employed,NaN,2.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers;When I want to...,All skills. AI is a flop.,104413.0,9.0
2,3,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Independent contractor, freelancer, or self-em...",None of the above,10.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code;GitHub Copilot;Google Gemini,NaN,When I don’t trust AI’s answers;When I want to...,"Understand how things actually work, problem s...",53061.0,8.0
3,4,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,None of the above,4.0,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code,NaN,When I don’t trust AI’s answers;When I want to...,NaN,36197.0,6.0
4,5,I am a developer by profession,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)","Independent contractor, freelancer, or self-em...","Caring for dependents (children, elderly, etc.)",21.0,"No, I am not new to coding and did not learn n...",NaN,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers,"critical thinking, the skill to define the tas...",60000.0,7.0


In [5]:
df.tail()

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,...,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
49186,49187,I code primarily as a hobby,18-24 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
49187,49188,I am a developer by profession,45-54 years old,Some college/university study without earning ...,Employed,"Caring for dependents (children, elderly, etc.)",25.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools req...",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
49188,49189,I am a developer by profession,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,None of the above,17.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,9.0
49189,49190,I am a developer by profession,25-34 years old,"Professional degree (JD, MD, Ph.D, Ed.D, etc.)","Independent contractor, freelancer, or self-em...",None of the above,2.0,"Yes, I am not new to coding but am learning ne...",Videos (not associated with specific online co...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
49190,49191,"I am not primarily a developer, but I write co...",18-24 years old,Some college/university study without earning ...,Student,Attending school (full-time),NaN,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [6]:
# Tipos
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 49191 entries, 0 to 49190
Columns: 172 entries, ResponseId to JobSat
dtypes: float64(52), int64(1), str(119)
memory usage: 64.6 MB


In [7]:
# Estadísticos básicos
df.describe(include="object")

C:\Users\maxim\AppData\Local\Temp\ipykernel_34104\1697761023.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")


,MainBranch,Age,EdLevel,Employment,EmploymentAddl,LearnCodeChoose,LearnCode,LearnCodeAI,AILearnHow,DevType,...,AIAgentKnowledge,AIAgentKnowWrite,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen
count,49191,49191,48149,48339,44875,46858,33556,45201,28257,43680,...,3401,766,3760,478,2692,264,8332,859,29194,22540
unique,6,7,8,6,171,3,2767,5,1677,32,...,528,561,620,332,314,195,1311,421,574,20138
top,I am a developer by profession,25-34 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,None of the above,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools req...",AI CodeGen tools or AI-enabled apps,"Developer, full-stack",...,GitHub MCP Server,Copilot,Ollama,n8n,Grafana + Prometheus,no,ChatGPT;GitHub Copilot,Cursor,When I don’t trust AI’s answers;When I want to...,Problem solving
freq,37467,16519,20278,33750,18613,32362,632,16403,2555,12351,...,657,15,796,28,588,13,669,106,3050,133


Se trata de un dataset completo y con gran cantidad de datos. No obstante, sería imprescindible escoger las columnas y filas necesarias para el EDA antes de continuar, para facilitar la exploración. Es decir, trabajar directamente sobre los datos que interesan

### Selección de columnas y filtrado de filas

Como realmente el estudio requiere datos relacionados con el perfil profesional en data y las hipotesis expuestas, entonces se elimina todas las columnas y filas que no tengan relación. Así se podrá comprobar el estado de nuestros datos de manera objetiva y eficiente.

Se expande la visión de todas las columnas del DF creado para visualizar los datos que contienen cada una de las columnas, posteriormente se escogerá las más adecuadas para las hipótesis plantadas.

In [8]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', 50)
df.head(5)

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,AILearnHow,YearsCode,DevType,OrgSize,ICorPM,RemoteWork,PurchaseInfluence,TechEndorseIntro,TechEndorse_1,TechEndorse_2,TechEndorse_3,TechEndorse_4,TechEndorse_5,TechEndorse_6,TechEndorse_7,TechEndorse_8,TechEndorse_9,TechEndorse_13,TechEndorse_13_TEXT,TechOppose_1,TechOppose_2,TechOppose_3,TechOppose_5,TechOppose_7,TechOppose_9,TechOppose_11,TechOppose_13,TechOppose_16,TechOppose_15,TechOppose_15_TEXT,Industry,JobSatPoints_1,JobSatPoints_2,JobSatPoints_3,JobSatPoints_4,JobSatPoints_5,JobSatPoints_6,JobSatPoints_7,JobSatPoints_8,JobSatPoints_9,JobSatPoints_10,JobSatPoints_11,JobSatPoints_13,JobSatPoints_14,JobSatPoints_15,JobSatPoints_16,JobSatPoints_15_TEXT,AIThreat,NewRole,ToolCountWork,ToolCountPersonal,Country,Currency,CompTotal,LanguageChoice,LanguageHaveWorkedWith,LanguageWantToWorkWith,LanguageAdmired,LanguagesHaveEntry,LanguagesWantEntry,DatabaseChoice,DatabaseHaveWorkedWith,DatabaseWantToWorkWith,DatabaseAdmired,DatabaseHaveEntry,DatabaseWantEntry,PlatformChoice,PlatformHaveWorkedWith,PlatformWantToWorkWith,PlatformAdmired,PlatformHaveEntry,PlatformWantEntry,WebframeChoice,WebframeHaveWorkedWith,WebframeWantToWorkWith,WebframeAdmired,WebframeHaveEntry,WebframeWantEntry,DevEnvsChoice,DevEnvsHaveWorkedWith,DevEnvsWantToWorkWith,DevEnvsAdmired,DevEnvHaveEntry,DevEnvWantEntry,SOTagsHaveWorkedWith,SOTagsWantToWorkWith,SOTagsAdmired,SOTagsHaveEntry,SOTagsWant Entry,OpSysPersonal use,OpSysProfessional use,OfficeStackAsyncHaveWorkedWith,OfficeStackAsyncWantToWorkWith,OfficeStackAsyncAdmired,OfficeStackHaveEntry,OfficeStackWantEntry,CommPlatformHaveWorkedWith,CommPlatformWantToWorkWith,CommPlatformAdmired,CommPlatformHaveEntr,CommPlatformWantEntr,AIModelsChoice,AIModelsHaveWorkedWith,AIModelsWantToWorkWith,AIModelsAdmired,AIModelsHaveEntry,AIModelsWantEntry,SOAccount,SOVisitFreq,SODuration,SOPartFreq,SO_Dev_Content,SO_Actions_1,SO_Actions_16,SO_Actions_3,SO_Actions_4,SO_Actions_5,SO_Actions_6,SO_Actions_9,SO_Actions_7,SO_Actions_10,SO_Actions_15,SO_Actions_15_TEXT,SOComm,SOFriction,AISelect,AISent,AIAcc,AIComplex,AIToolCurrently partially AI,AIToolDon't plan to use AI for this task,AIToolPlan to partially use AI,AIToolPlan to mostly use AI,AIToolCurrently mostly AI,AIFrustration,AIExplain,AIAgents,AIAgentChange,AIAgent_Uses,AgentUsesGeneral,AIAgentImpactSomewhat agree,AIAgentImpactNeutral,AIAgentImpactSomewhat disagree,AIAgentImpactStrongly agree,AIAgentImpactStrongly disagree,AIAgentChallengesNeutral,AIAgentChallengesSomewhat disagree,AIAgentChallengesStrongly agree,AIAgentChallengesSomewhat agree,AIAgentChallengesStrongly disagree,AIAgentKnowledge,AIAgentKnowWrite,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",AI CodeGen tools or AI-enabled apps,14.0,"Developer, mobile",20 to 99 employees,People manager,Remote,"Yes, I influenced the purchase of a substantia...",Work,10.0,7.0,9.0,6.0,3.0,11.0,12.0,1.0,8.0,14.0,NaN,15.0,7.0,8.0,12.0,11.0,1.0,6.0,13.0,3.0,16.0,NaN,Fintech,3.0,1.0,4.0,9.0,5.0,10.0,12.0,11.0,2.0,6.0,7.0,13.0,14.0,15.0,8.0,NaN,I'm not sure,I have neither consider or transitioned into a...,7.0,3.0,Ukraine,EUR European Euro,52800.0,Yes,Bash/Shell (all shells);Dart;SQL,Dart,Dart,NaN,NaN,Yes,Cloud Firestore;PostgreSQL,NaN,NaN,NaN,NaN,Yes,Amazon Web Services (AWS);Cloudflare;Firebase;...,NaN,NaN,NaN,NaN,No,NaN,NaN,NaN,NaN,NaN,Yes,Android Studio;Notepad++;Visual Studio;Visual ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Windows;MacOS;Android,Windows;MacOS;Android;iOS;iPadOS,Confluence;GitHub;GitLab;Jira;Markdo

Los datos más interesantes se encuentran en la siguientes columnas, las cuales serán seleccionadas para el DataFrame final con el cual se trabajará

In [9]:
columnas_interes = [
    "ConvertedCompYearly", # Variable fundamental para las hipotesis; Salario anual en dolares
    "Age", # Importante para H1: Cuanta mayor edad mayor salario
    "EdLevel", # Importante para H2: Cuanta mayor cualificación académica mayor salario
    "WorkExp", # Importante para H3: Experiencia laboral
    "YearsCode", # Importante para H3: Experiencia profesional 
    "LanguageHaveWorkedWith", # Importante para H3: Lenguajes de programación que dominan
    "Employment", # Importante para H4: Los trabajadores por cuenta ajena tienen una mayor renta    
    "AISelect", # Importante para H5: Trabajar con IA te ofrece mayor oportunidad laboral.
    "RemoteWork", # Trabajo remoto, híbrido o presencial (?)
    "Industry", # Tipo de industria, puede variar ciertos factores, ya sea por la demanda u oferta
    "Country", # País donde ejerce su profesión. Puede variar las condiciones según el país.
    "JobSat", # Satisfacción en el trabajo. Se puede añadir como variable extra para analizar como resultado además del salario.
    "DevType", # Tipo de desarrollador, se filtrará, posteriormente, por los relacionados con la Ciencia de Datos
]

# A considerar o profundizar: LanguageWantToWorkWith	LanguageAdmired 
# AIModelsChoice	AIModelsHaveWorkedWith	AIModelsWantToWorkWith	AIModelsAdmired	
# DatabaseHaveWorkedWith	DatabaseWantToWorkWith	DatabaseAdmired 
# PlatformHaveWorkedWith	PlatformWantToWorkWith	PlatformAdmired
# OfficeStackAsyncHaveWorkedWith	OfficeStackAsyncWantToWorkWith	OfficeStackAsyncAdmired

In [10]:
df_eda = df[columnas_interes]
print(df_eda.shape)
df_eda.head()

(49191, 13)


,ConvertedCompYearly,Age,EdLevel,WorkExp,YearsCode,LanguageHaveWorkedWith,Employment,AISelect,RemoteWork,Industry,Country,JobSat,DevType
0,61256.0,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",8.0,14.0,Bash/Shell (all shells);Dart;SQL,Employed,"Yes, I use AI tools monthly or infrequently",Remote,Fintech,Ukraine,10.0,"Developer, mobile"
1,104413.0,25-34 years old,"Associate degree (A.A., A.S., etc.)",2.0,10.0,Java,Employed,"Yes, I use AI tools weekly","Hybrid (some in-person, leans heavy to flexibi...",Retail and Consumer Services,Netherlands,9.0,"Developer, back-end"
2,53061.0,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",10.0,12.0,Dart;HTML/CSS;JavaScript;TypeScript,"Independent contractor, freelancer, or self-em...","Yes, I use AI tools daily",NaN,Software Development,Ukraine,8.0,"Developer, front-end"
3,36197.0,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",4.0,5.0,Java;Kotlin;SQL,Employed,"Yes, I use AI tools weekly",Remote,Retail and Consumer Services,Ukraine,6.0,"Developer, back-end"
4,60000.0,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",21.0,22.0,C;C#;C++;Delphi;HTML/CSS;Java;JavaScript;Lua;P...,"Independent contractor, freelancer, or self-em...","Yes, I use AI tools weekly",NaN,Software Development,Ukraine,7.0,Engineering manager


Obtenido el DF con las columnas de interes para el estudio, se procede a visualizar cuales son los datos que contiene la columna "DevType" (Tipo de Desarrollador) y posteriormente, filtrar por los que tengan relación en el sector data.

In [11]:
df_eda["DevType"].unique()

<StringArray>
[                            'Developer, mobile',
                           'Developer, back-end',
                          'Developer, front-end',
                           'Engineering manager',
                         'Developer, full-stack',
              'Architect, software or solutions',
                                 'Data engineer',
                 'Financial analyst or engineer',
                           'Academic researcher',
   'Developer, embedded applications or devices',
         'Cybersecurity or InfoSec professional',
                   'Developer, game or graphics',
                                'AI/ML engineer',
          'Senior executive (C-suite, VP, etc.)',
 'Developer, desktop or enterprise applications',
                       'Other (please specify):',
              'Founder, technology or otherwise',
                                       'Retired',
             'Developer, AI apps or physical AI',
               'DevOps engineer or p

In [12]:
roles_data = [
    "Data engineer",
    "Data scientist",
    "Data or business analyst",
    "Applied scientist",
    "Database administrator or engineer",    
    "AI/ML engineer",
    "Developer, AI apps or physical AI"
]

# A considerar: Financial analyst or engineer (es algo más especializado, descartado, al igual que otros perfiles como arquitectos de BBDD, etc).

In [13]:
patron = "|".join(roles_data)

df_data = df_eda[df_eda["DevType"].str.contains(patron, case=False, na=False)].copy()

print(f"Filas totales:    {len(df_eda)}")
print(f"Filas data roles: {len(df_data)}")
print(f"Porcentaje:       {len(df_data)/len(df_eda)*100}%")

Filas totales:    49191
Filas data roles: 3080
Porcentaje:       6.26130796283873%


Se ha filtrado de las 49 mil filas a 3080 filas, es decir, tenemos un 6.26% aproximadamente del dataset principal.

Como se puede observar, este primer proceso ahorra mucho tiempo y trabajo, para poder centrar todos los siguientes pasos exclusivamente en el objetivo principal.

### Comprobación de calidad del subconjunto

Se obtiene la vision general y calidad de los datos del dataset.

In [14]:
# Resumen rápido de calidad
calidad_datos(df_data)

,dtype,nulos,pct_nulos,unicos
ConvertedCompYearly,float64,1334,43.31,921
Age,str,0,0.00,7
EdLevel,str,14,0.45,8
WorkExp,float64,204,6.62,50
YearsCode,float64,124,4.03,62
LanguageHaveWorkedWith,str,907,29.45,1332
Employment,str,0,0.00,6
AISelect,str,760,24.68,5
RemoteWork,str,579,18.80,5
Industry,str,609,19.77,15


Información a destacar:
- La columna de "Age" es de tipo string debido a que sus datos son rangos de edad. Se decide mantenerlo como str, y no cambiar a int o float, debido a que los datos no son específicos. En cuanto al resto de tipos de datos, es adecuado.
- Se puede comprobar que existen valores nulos, repartidos entre diferentes columnas. A continuación se tratarán.
- En cuanto a los valores únicos, existe mucha variedad la cual se analizará próximamente.

### Limpieza

In [15]:
# Duplicados
df_data.duplicated().sum()

np.int64(4)

In [16]:
df_data.drop_duplicates(inplace=True)
df_data.duplicated().sum()

np.int64(0)

In [17]:
# Eliminar columnas constantes (una sola categoría)
cols_constantes = [c for c in df_data.columns if df_data[c].nunique() == 1]
df_data.drop(columns=cols_constantes, inplace=True)

In [18]:
# Missings por columna
df_data.isnull().sum().sort_values(ascending=False)

JobSat                    1435
ConvertedCompYearly       1330
LanguageHaveWorkedWith     903
AISelect                   757
Industry                   605
Country                    591
RemoteWork                 578
WorkExp                    203
YearsCode                  123
EdLevel                     14
Age                          0
Employment                   0
DevType                      0
dtype: int64

In [19]:
serie_nulos = df_data.isna().mean() * 100
serie_nulos

ConvertedCompYearly       43.237971
Age                        0.000000
EdLevel                    0.455137
WorkExp                    6.599480
YearsCode                  3.998700
LanguageHaveWorkedWith    29.356307
Employment                 0.000000
AISelect                  24.609883
RemoteWork                18.790637
Industry                  19.668401
Country                   19.213264
JobSat                    46.651495
DevType                    0.000000
dtype: float64

La columna JobSat, se deja como NaN, debido a que es una respuesta valida. Además, de ser una variable que interfiere directamente en las hipótesis planteadas y contiene demasiados valores NaN como para eliminarlos, se toma la decisión de mantenerlo tal y como está.

Luego, la columna más importante del dataset tiene un 43.23% de valores nulos, se podría sustituir por la media según el país y el tipo de desarrollador, sin embargo es una información que sería sesgada finalmente y no enriquecería el estudio, y como existe un margen de datos relativamente amplio, es por lo que se decide eliminar aquellas filas nulas que tan importantes son para el estudio.

In [20]:
# Eliminar filas con valores nulos en "ConvertedCompYearly"
df_data.dropna(subset=["ConvertedCompYearly"], inplace=True)

Aplicamos mismo método para EdLevel, debido a que son pocas filas las que tienen valor nulo, no interfiere en el estudio.

In [21]:
df_data = df_data.dropna(subset=["EdLevel"])

RemoteWork realmente no aporta nada importante para el estudio, se elimina la columna

In [22]:
# Eliminar columna RemoteWork
df_data.drop(columns=["RemoteWork"], inplace= True)

Lo mismo ocurre con el lenguaje de programación, no aportaría nada para este estudio en concreto, se elimina

In [23]:
df_data.drop(columns=["LanguageHaveWorkedWith"], inplace= True)

Los datos de la columna AISelect pueden resumirse en 3; Sí, No, NC (No Contesta)

In [24]:
df_data["AISelect"].value_counts()

AISelect
Yes, I use AI tools daily                      895
Yes, I use AI tools weekly                     275
No, and I don't plan to                        205
Yes, I use AI tools monthly or infrequently    191
No, but I plan to soon                         103
Name: count, dtype: int64

In [25]:
mapeo_ai = {
    "Yes, I use AI tools daily": "Yes",
    "Yes, I use AI tools weekly": "Yes",
    "Yes, I use AI tools monthly or infrequently": "Yes",
    "No, and I don't plan to": "No",
    "No, but I plan to soon": "No"
}

df_data["AISelect"] = df_data["AISelect"].replace(mapeo_ai).fillna("NA") # NA : No Answer

In [26]:
mapeo_edlevel = {
    "Bachelor’s degree (B.A., B.S., B.Eng., etc.)": "Bachelor",
    "Professional degree (JD, MD, Ph.D, Ed.D, etc.)": "Professional",
    "Master’s degree (M.A., M.S., M.Eng., MBA, etc.)": "Master",
    "Some college/university study without earning a degree": "Some College",
    "Associate degree (A.A., A.S., etc.)": "Associate",
    "Secondary school (e.g. American high school, German Realschule or Gymnasium, etc.)": "Secondary",
    "Primary/elementary school": "Primary",
    "Other (please specify:)": "Other"
}

df_data["EdLevel"] = df_data["EdLevel"].replace(mapeo_edlevel)

In [ ]:
mapeo_employment = {                                            
    'Freelancer':'Freelancer'
    }

df_data["Employment"] = df_data["Employment"].replace(mapeo_employment)

In [28]:
mapeo_devtype = {                                            
    "Data or business analyst":"Data analyst",
    "Database administrator or engineer":"Database administrator",    
    "Developer, AI apps or physical AI":"AI developer"
    }

df_data["DevType"] = df_data["DevType"].replace(mapeo_devtype)

In [29]:
mapeo_country = {                                            
    'United Kingdom of Great Britain and Northern Ireland':'UK and Ireland',
    'United States of America':'USA',
    'Iran, Islamic Republic of...':'Iran',
    'Hong Kong (S.A.R.)':'Hong Kong',
    'United Arab Emirates':'UAE',
    'Russian Federation':'Russia',
    "Côte d'Ivoire":'Ivory Coast',
    'Congo, Republic of the...':'Congo'
    }

df_data["Country"] = df_data["Country"].replace(mapeo_country)

Las tres columnas que quedan, se aplicará la moda y media, respectivamente. Antes se indica en una nueva columna que era un valor nulo en esa columna y fila.

In [30]:
# Marcar antes de imputar
df_data["fue_missing_YearCode"] = df_data["YearsCode"].isna()
df_data["fue_missing_WorkExp"] = df_data["WorkExp"].isna()
df_data["fue_missing_Industry"] = df_data["Industry"].isna()

In [31]:
df_data["Industry"] = df_data["Industry"].fillna(df_data["Industry"].mode()[0])

In [32]:
df_data["WorkExp"] = df_data["WorkExp"].fillna(df_data["WorkExp"].median())

In [33]:
df_data["YearsCode"] = df_data["YearsCode"].fillna(df_data["YearsCode"].median())

Desde el análisis univariante se ha comprobado que existen outliers, de los cuales el valor 100 se descarta. 

In [34]:
df_data["YearsCode"].unique()

array([ 15.,  26.,  23.,  20.,  42.,  10.,   8.,  12.,  18.,  29.,  25.,
         3.,  35.,  38.,   9.,  31.,  24.,  16.,  19.,  27.,  30.,  43.,
        17.,  11.,  28.,   5.,  34.,   4.,  14.,  44.,   7.,  39.,  33.,
        40.,  45.,  41.,   6.,  13.,  58.,  37.,  46.,  22.,  50.,   2.,
        36.,  49.,   1.,  21.,  32.,  52.,  60.,  48., 100.,  53.,  57.,
        55.,  47.])

In [35]:
df_data.loc[df_data["YearsCode"] == 100, "YearsCode"] = pd.NA

In [36]:
df_data = df_data.dropna(subset=["YearsCode"])

Por último, volvemos a comprobar

In [37]:
df_data.isnull().sum().sort_values(ascending=False)

JobSat                  552
ConvertedCompYearly       0
EdLevel                   0
WorkExp                   0
YearsCode                 0
Age                       0
Employment                0
AISelect                  0
Industry                  0
Country                   0
DevType                   0
fue_missing_YearCode      0
fue_missing_WorkExp       0
fue_missing_Industry      0
dtype: int64

In [38]:
df_data.info()

<class 'pandas.DataFrame'>
Index: 1741 entries, 8 to 49082
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   ConvertedCompYearly   1741 non-null   float64
 1   Age                   1741 non-null   str    
 2   EdLevel               1741 non-null   str    
 3   WorkExp               1741 non-null   float64
 4   YearsCode             1741 non-null   float64
 5   Employment            1741 non-null   str    
 6   AISelect              1741 non-null   str    
 7   Industry              1741 non-null   str    
 8   Country               1741 non-null   str    
 9   JobSat                1189 non-null   float64
 10  DevType               1741 non-null   str    
 11  fue_missing_YearCode  1741 non-null   bool   
 12  fue_missing_WorkExp   1741 non-null   bool   
 13  fue_missing_Industry  1741 non-null   bool   
dtypes: bool(3), float64(4), str(7)
memory usage: 168.3 KB


### Data dictionary del dataset limpio

Se documenta qué es cada variable, su tipo, si es continua/categórica, y qué crees que va a aportar. 

In [39]:
descripciones = {
    "ConvertedCompYearly": "Salario anual convertido a USD",
    "Country": "País de residencia del encuestado",
    "YearsCodePro": "Años de experiencia programando profesionalmente",
    "EdLevel": "Nivel educativo más alto alcanzado",
    "DevType": "Tipo de rol o puesto de trabajo",
    "OrgSize": "Tamaño de la organización donde trabaja",
    "AISelect": "Uso de herramientas de IA en el trabajo",
    "JobSat": "Nivel de satisfacción laboral",
    "Industry": "Sector o industria de la empresa",
    "WorkExp": "Años de experiencia laboral total",
    "YearsCode": "Años totales programando (incluye antes de trabajar)"
}

data_dict = pd.DataFrame({
    "columna": df_data.columns,
    "descripcion_es": [descripciones.get(c, "Sin descripción") for c in df_data.columns],
    "dtype": df_data.dtypes.values,
    "nulos_pct": (df_data.isnull().mean() * 100).round(1).values,
    "unicos": df_data.nunique().values,
    "ejemplo": [df_data[c].dropna().iloc[0] if len(df_data[c].dropna()) > 0 else None for c in df_data.columns]
})

data_dict.to_csv("../data/data_dictionary.csv", index=False)
data_dict

,columna,descripcion_es,dtype,nulos_pct,unicos,ejemplo
0,ConvertedCompYearly,Salario anual convertido a USD,float64,0.0,919,70000.0
1,Age,Sin descripción,str,0.0,7,25-34 years old
2,EdLevel,Nivel educativo más alto alcanzado,str,0.0,8,Bachelor
3,WorkExp,Años de experiencia laboral total,float64,0.0,48,9.0
4,YearsCode,Años totales programando (incluye antes de tra...,float64,0.0,56,15.0
5,Employment,Sin descripción,str,0.0,6,Employed
6,AISelect,Uso de herramientas de IA en el trabajo,str,0.0,3,Yes
7,Industry,Sector o industria de la empresa,str,0.0,15,Banking/Financial Services
8,Country,País de residencia del encuestado,str,0.0,104,Ukraine
9,JobSat,Nivel de satisfacción laboral,float64,31.7,11,8.0


### Guardar CSV limpio

In [40]:
# Generamos copia de csv limpio para EDA
df_data.to_csv("../data/develop_dataset_2025_limpio.csv")